### PR TITLE
bpo-31942: Document optional support of start and stop attributes in Sequence.index method

### PR DESCRIPTION
**Pull request by nitishch**
I changed the method docstring and also the html documentation to say that not all implementations are required to support  start and stop attributes. Is a NEWS entry required for this?

Bug: https://bugs.python.org/issue31942

https://bugs.python.org/issue31942

### CODE REVIEW

**Comment by vstinner** [comment]
@serhiy-storchaka: Would you mind to review this PR, since you worked on that topic and created https://bugs.python.org/issue31942 ?

File: Lib/_collections_abc.py [hunk 1]
```diff
@@ -899,6 +899,10 @@ def __reversed__(self):
     def index(self, value, start=0, stop=None):
         '''S.index(value, [start, [stop]]) -> integer -- return first index of value.
            Raises ValueError if the value is not present.
+
+           It is not required that all concrete implementations of
+           the Sequence class support start and stop attributes.
```
Comment (vstinner) [inline Lib/_collections_abc.py]: I would rephrase it as:

Supporting start and stop optional arguments is optional, but recommended.

File: Doc/library/stdtypes.rst [hunk 2]
```diff
@@ -973,11 +973,12 @@ Notes:
 
 (8)
    ``index`` raises :exc:`ValueError` when *x* is not found in *s*.
-   When supported, the additional arguments to the index method allow
-   efficient searching of subsections of the sequence. Passing the extra
-   arguments is roughly equivalent to using ``s[i:j].index(x)``, only
-   without copying any data and with the returned index being relative to
-   the start of the sequence rather than the start of the slice.
+   Not all implementations support passing the additional arguments *i* and *j*.
+   When supported, these arguments allow efficient searching of
```
Comment (vstinner) [inline Doc/library/stdtypes.rst]: "Not all implementations support passing the additional arguments *i* and *j*." seems redundant with "When supported".

Comment (nitishch) [reply Doc/library/stdtypes.rst]: Done.

File: Lib/_collections_abc.py (hunk 1 above)
Comment (nitishch) [reply Lib/_collections_abc.py]: I changed it to 'Supporting start and stop arguments is optional, but recommended.". I removed the first 'optional' word because the senses of both 'optional' words is different and might confuse the reader and also the sense of first 'optional' is conveyed by the function signature anyway.

**Review by vstinner — APPROVED** [review]
LGTM.

File: Doc/library/stdtypes.rst [hunk 3]
```diff
@@ -973,11 +973,12 @@ Notes:
...
 (8)
    ``index`` raises :exc:`ValueError` when *x* is not found in *s*.
-   When supported, the additional arguments to the index method allow
-   efficient searching of subsections of the sequence. Passing the extra
-   arguments is roughly equivalent to using ``s[i:j].index(x)``, only
-   without copying any data and with the returned index being relative to
-   the start of the sequence rather than the start of the slice.
+   Not all implementations support passing the additional arguments
+   *i* and *j*.  These arguments allow efficient searching of
+   subsections of the sequence. Passing the extra arguments is roughly
```
Comment (serhiy-storchaka) [inline Doc/library/stdtypes.rst]: Avoid reflowing the lines that are not changed.

File: Lib/_collections_abc.py [hunk 4]
```diff
@@ -899,6 +899,10 @@ def __reversed__(self):
     def index(self, value, start=0, stop=None):
         '''S.index(value, [start, [stop]]) -> integer -- return first index of value.
            Raises ValueError if the value is not present.
+
+           Supporting start and stop arguments is optional, but
+           recommended.
+
```
Comment (serhiy-storchaka) [inline Lib/_collections_abc.py]: Redundant empty line.

File: Doc/library/stdtypes.rst (hunk 3 above)
Comment (nitishch) [reply Doc/library/stdtypes.rst]: Minimised the reflowing.

File: Lib/_collections_abc.py (hunk 4 above)
Comment (nitishch) [reply Lib/_collections_abc.py]: Removed it.

**Review by serhiy-storchaka — APPROVED** [review]
